test/test_sandbox_app.py

```diff
@@ -0,0 +1,148 @@
+DEFAULT_CFG_OPTS = ["habitat.dataset.split=minival"]
+
+SandboxDriverArgs = namedtuple(
+    "SandboxDriverArgs",
+    [
+        "cfg",
+        "opts",
+        "width",
+        "height",
+        "walk_pose_path",
+        "never_end",
+        "disable_inverse_kinematics",
+        "humanoid_user_agent",
+        "use_batch_renderer",
+    ],
+)
+
+
+@pytest.mark.skipif(
+    not habitat_sim.built_with_bullet,
+    reason="Bullet physics used for validation.",
+)
+@pytest.mark.skipif(
+    not osp.exists("data/replica_cad"),
+    reason="Requires the replica_cad",
+)
+@pytest.mark.skipif(
+    not osp.exists("data/datasets/rearrange_pick/replica_cad/v0"),
+    reason="Requires the replica_cad v0 rearrange pick dataset",
+)
+@pytest.mark.parametrize(
+    "args",
+    [
+        SandboxDriverArgs(
+            cfg=DEFAULT_CFG,
+            opts=DEFAULT_CFG_OPTS,
+            width=DEFAULT_APP_WINDOW_WIDTH,
+            height=DEFAULT_APP_WINDOW_HEIGHT,
+            walk_pose_path=DEFAULT_POSE_PATH,
+            never_end=True,
+            disable_inverse_kinematics=True,
+            humanoid_user_agent=True,
+            use_batch_renderer=False,
+        )
+    ],
+)
+def test_sandbox_driver(args):
+    # get config
+    config = habitat.get_config(args.cfg, args.opts)
+    config = update_habitat_config(config, args)
+
+    glfw_config = Application.Configuration()
+    glfw_config.title = "Sandbox App"
+    glfw_config.size = (args.width, args.height)
+    gui_app_wrapper = GuiApplication(glfw_config, 30)
+    framebuffer_size = gui_app_wrapper.get_framebuffer_size()
+
+    # init SandboxDriver
+    sim_input = GuiInput()
+    driver = SandboxDriver(args, config, gui_input=sim_input)
+    app_renderer = ReplayGuiAppRenderer(
+        framebuffer_size.x, framebuffer_size.y, args.use_batch_renderer
+    )
+    driver.set_debug_line_render(
+        app_renderer._replay_renderer.debug_line_render(0)
+    )
+
+    # assert post_sim_update_dict match expected result
+    post_sim_update_dict = driver.sim_update(
+        dt=None
+    )  # dt argument is not used within sim_update
+    # print(post_sim_update_dict)
+
+    # todo: add asserts here
+
+    # post_sim_update
+    app_renderer.post_sim_update(post_sim_update_dict)
+
+    # render keyframe to the MutableImageView2D
+    # 1) doesn't work:
+    buffer = np.empty(
+        (
+            framebuffer_size.y,
+            framebuffer_size.x,
+            3,
+        ),
+        dtype=np.uint8,
+    )
+    mutable_image_view = mn.MutableImageView2D(
+        mn.PixelFormat.RGB8_UNORM, framebuffer_size, buffer
+    )
+    # app_renderer._replay_renderer.set_sensor_transform(
+    #     0, app_renderer._sensor_uuid, app_renderer.cam_transform
+    # )
+    # AssertionError: ESP_CHECK failed: Renderer was not created with a background render thread, cannot do async drawing
+    # app_renderer._replay_renderer.render([mutable_image_view])
```

Comment:
Once you switch to a standalone replay renderer, you may still his this same error. @0mdc should weigh in as he created these renderers and knows about the various `render` methods.

Reply:
A few things could be happening here.

1. Make sure that you are using an up-to-date habitat-sim. This error was occurring prior to [this fix](https://github.com/facebookresearch/habitat-sim/pull/2043).
2. `ReplayGuiAppRenderer` is windowed, hence it already has a context. It tries to create a standalone context here. Therefore, it fails. The following solutions could fix this:
    * Creating path so that `ReplayGuiAppRenderer` can be run in headless mode.
    * I can add a python binding to copy the framebuffer into an `MutableImageView`. This already exists in the backend. In this case, you would have to call `replay_renderer.render(framebuffer)`, then something like `replay_renderer.copy_color_into(image_views)`.